server/src/test/java/org/opensearch/indices/replication/checkpoint/PublishReferencedSegmentsActionTests.java

```diff
@@ -0,0 +1,230 @@
+        when(recoverySettings.getMergedSegmentReplicationTimeout()).thenReturn(new TimeValue(1000));
+        when(shardRouting.allocationId()).thenReturn(allocationId);
+        when(allocationId.getId()).thenReturn("1");
+        when(recoveryState.getTargetNode()).thenReturn(clusterService.localNode());
+        when(indexShard.routingEntry()).thenReturn(shardRouting);
+        when(indexShard.getPendingPrimaryTerm()).thenReturn(1L);
+        when(indexShard.recoveryState()).thenReturn(recoveryState);
+        when(indexShard.getRecoverySettings()).thenReturn(recoverySettings);
+
+        final PublishReferencedSegmentsAction action = new PublishReferencedSegmentsAction(
+            Settings.EMPTY,
+            transportService,
+            clusterService,
+            indicesService,
+            threadPool,
+            shardStateAction,
+            new ActionFilters(Collections.emptySet())
+        );
+
+        final ReferencedSegmentsCheckpoint checkpoint = new ReferencedSegmentsCheckpoint(
+            indexShard.shardId(),
+            1,
+            1,
+            1111,
+            Codec.getDefault().getName(),
+            Collections.emptyMap(),
+            Sets.newHashSet("_1", "_2", "_3")
+        );
+
+        action.publish(indexShard, checkpoint);
+    }
+
+    public void testPublishReferencedSegmentsOnPrimary() {
+        final IndicesService indicesService = mock(IndicesService.class);
+
+        final Index index = new Index("index", "uuid");
+        final IndexService indexService = mock(IndexService.class);
+        when(indicesService.indexServiceSafe(index)).thenReturn(indexService);
+
+        final int id = randomIntBetween(0, 4);
+        final IndexShard indexShard = mock(IndexShard.class);
+        when(indexService.getShard(id)).thenReturn(indexShard);
+
+        final ShardId shardId = new ShardId(index, id);
+        when(indexShard.shardId()).thenReturn(shardId);
+
+        final PublishReferencedSegmentsAction action = new PublishReferencedSegmentsAction(
+            Settings.EMPTY,
+            transportService,
+            clusterService,
+            indicesService,
+            threadPool,
+            shardStateAction,
+            new ActionFilters(Collections.emptySet())
+        );
+
+        final ReferencedSegmentsCheckpoint checkpoint = new ReferencedSegmentsCheckpoint(
+            indexShard.shardId(),
+            1,
+            1,
+            1111,
+            Codec.getDefault().getName(),
+            Collections.emptyMap(),
+            Sets.newHashSet("_1", "_2", "_3")
+        );
+        final PublishReferencedSegmentsRequest request = new PublishReferencedSegmentsRequest(checkpoint);
+
+        action.shardOperationOnPrimary(request, indexShard, ActionTestUtils.assertNoFailureListener(result -> {
+            // we should forward the request containing the current publish checkpoint to the replica
+            assertThat(result.replicaRequest(), sameInstance(request));
```

Comment:
should we have some counter here that is validated outside to ensure that this block of code was definitely executed?